### ISSUE TITLE
named backreference failing

### ISSUE THREAD

**Issue by kameamea** [slevithan/xregexp#68]
Running this code with firefox 29 and xregexp 3.0pre shows true for both alerts.
The \\k<sep> doesn't work as expected and as \\2 in dateOK does. If we remove the
"| \\s* (?<fullmonth> August ) " -part then the bug doesn't show up anymore.

```
var dateOK = XRegExp.build(' \
( \
 (?<day> [0-3]?\\d) \
\\s* ((?<sep> [./\\s]) ) \\s*  \
(?<month> (1[012]|0?\\d))  \
| \\s* (?<fullmonth> August ) )  \
( \\s* \\2 \\s*  \
(?<year> (20)?[012]\\d)  \
)? ',{},'xni');

var dateBUG = XRegExp.build(' \
( \
 (?<day> [0-3]?\\d) \
\\s* ((?<sep> [./\\s]) ) \\s*  \
(?<month> (1[012]|0?\\d))  \
| \\s* (?<fullmonth> August ) )  \
( \\s* \\k<sep> \\s*  \
(?<year> (20)?[012]\\d)  \
)? ',{},'xni');

alert(XRegExp.exec("05/07 09",dateOK).year == "");
alert(XRegExp.exec("05/07 09",dateBUG).year == "09");
```

**Comment by slevithan** [comment]
Running in Firefox 40.0.2 and Chrome 44, I get `false` and `true` (not `true` for both). Can you clarify if you're still able to repro this, and explain / make it easier to follow what the bug is, preferably with a reduced test case.

**Comment by kameamea** [comment]
I get a false and a true now as well, but that's only because instead of an empty string there is an "undefined" now, but the bug report is about something else. Admittedly I was not very clear when reporting this issue. I'm trying to explain it better now: The regex tries to detect a valid date, which might contain a year but can as well consist of day/month only. The separator between those can be a dot, a slash or a space. So:
```
var date1WithYear = "05/07/09";
var date2WithYear = "10.12.14";
var date1WithoutYear = "05/07 09";
```
has two valid dates with a year while the `date1WithoutYear` should be recognized as "05/07" only, the " 09" part is not belonging to the date anymore. To achieve this I use the `\\k<sep>`-Syntax. The `\\k` is supposed to check that the same separator is used before the "year"-part.  But:

```
alert(XRegExp.exec(date1WithYear,dateBUG).year == "09")
```
results in false. If I replace the  `\\k<sep>`-Syntax by `\\2` (this is the `dateOK`-regex), I get the expected result however:

```
alert(XRegExp.exec(date1WithYear,dateOK).year == "09")
```
delivers true.

In the original report I use `date1WithoutYear` which should detect no "year"-property but again dateBUG fails, dateOK works as expected. 

So perhaps I'm not using the `\\k` properly, otherwise I assume this is a bug that the named backreference is not working in this case. When I tried to strip down the example I couldn't reproduce it already after removing the "| \s* (? August ) " -part so I left it in its complex form.

**Comment by slevithan** [comment]
Okay, after spending some time closely analyzing your regexes and report I finally see that this is indeed a bug. It is stemming from the combination of `XRegExp.build`, flag `n`, and mixed named/numbered capture.

In the transpiled regex generated for `dateBUG`, `\k<sep>` incorrectly refers to backreference 4 (`\4`), and in `dateOK` you are of course referring directly to `\2`. If it wasn't for flag `n`, `\k<sep>` would indeed be referring to the fourth capturing group. And if you built these regexes using the standard `XRegExp` constructor rather than `XRegExp.build` (which you probably should since you are using no `.build` features here, just passing in an empty object as the second argument), this would also work correctly. It is something about the combination of `.build` and `n` that is causing this to generate the wrong result. There's probably some influence from other specifics of the regex as well, since you mentioned it was hard to produce a reduced test case.

This should be easier to investigate now that I at least partially understand the problem and have verified that there *is* a problem, which wasn't obvious to me before due to the regexes and explanation being hard to follow. Thanks so much for coming back to add more detail despite the long delay in getting back to you after you fixed your initial report.

To make it easier to compare your regexes, I've reformatted them as follows:

```
var dateOK = XRegExp.build('       \
  (                                \
    (?<day> [0-3]?\\d )            \
    \\s* ((?<sep> [./\\s] )) \\s*  \
    (?<month> (1[012]|0?\\d) )     \
  |                                \
    \\s* (?<fullmonth> August )    \
  )                                \
  (                                \
    \\s* \\2 \\s*                  \
    (?<year> (20)?[012]\\d )       \
  )?                               ',
{}, 'xni');

var dateBUG = XRegExp.build('      \
  (                                \
    (?<day> [0-3]?\\d )            \
    \\s* ((?<sep> [./\\s] )) \\s*  \
    (?<month> (1[012]|0?\\d) )     \
  |                                \
    \\s* (?<fullmonth> August )    \
  )                                \
  (                                \
    \\s* \\k<sep> \\s*             \
    (?<year> (20)?[012]\\d )       \
  )?                               ',
{}, 'xni');
```

**Comment by slevithan** [comment]
Maximally reduced test case:

```
XRegExp.build('((?<n>))\\k<n>', {}, 'n');
// Uncaught SyntaxError: Cannot use octal escape or backreference to undefined group \2
```

Note that this bug requires the use of both flag `n` and `XRegExp.build` (rather than just `XRegExp`) when creating the regex.
